### PR TITLE
Delete 0 or 1 instances on app level, or a single on instance level

### PR DESCRIPTION
**Pull request by jonmv**
@bratseth please review.

### CODE REVIEW

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/ApplicationController.java [hunk 1]
```diff
@@ -660,36 +662,56 @@ private DeployOptions withVersion(Version version, DeployOptions options) {
...
+        if (tenant.type() != Tenant.Type.user && credentials.isEmpty())
+            throw new IllegalArgumentException("Could not delete application '" + tenantName + "." + applicationName + "': No credentials provided");
+
+        // Find all instances of the application
+        List<ApplicationId> instances = asList(tenantName).stream()
+                                                          .map(Application::id)
+                                                          .filter(id -> id.application().equals(applicationName))
+                                                          .collect(Collectors.toList());
+        if (instances.size() > 1)
+            throw new IllegalArgumentException("Could not delete applicattion; more than one instance present: " + instances);
```
Comment (freva) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/ApplicationController.java]: `applicattion` -> `application`

**Review by freva — COMMENTED** [review]
If you delete the last instance, it should also delete the data in access control?

**Review by bratseth — APPROVED** [review]
LGTM

**Comment by jonmv** [comment]
I think the application (access control entries) should only be deleted when the application is deleted. The console should have two delete buttons — one for each instance, and one for the application itself. Disagree? @smorgrav @oyving @freva

**Comment by jonmv** [comment]
Actually, I'm not sure how much sense it makes to be able to delete single instances in the console, unless these are `dev` instances: instances in `prod` will be managed through `deployment.xml` ...

**Comment by jonmv** [comment]
We should revisit this when instance orchestration is done.

**Comment by freva** [comment]
> I think the application (access control entries) should only be deleted when the application is deleted. The console should have two delete buttons — one for each instance, and one for the application itself. Disagree? 

Agree, but until the controller actually has a concept of just the application, I think we need handle the case where you delete the last instance, because once you do that, controller will 404 anything under `/application/v4/tenant/<tenant>/application/<application>`. So we wont know in the console to show the delete application button.

**Pull request merged**